File: app/src/main/java/com/liweiyap/narradir/ui/ViewGroupSingleTargetSelector.java
package com.liweiyap.narradir.ui;

import android.view.View;
import android.view.ViewGroup;
import android.widget.Checkable;

public final class ViewGroupSingleTargetSelector
{
    private ViewGroupSingleTargetSelector(){}

    public static void addSingleTargetSelection(final ViewGroup viewGroup) throws RuntimeException
    {
        if (viewGroup == null)
        {
            return;
        }

        for (int childIdx = 0; childIdx < viewGroup.getChildCount(); ++childIdx)
        {
            View view = viewGroup.getChildAt(childIdx);

            if (!(view instanceof Checkable))
            {
                throw new RuntimeException(
                    "ViewGroupSingleTargetSelector::addSingleTargetSelection(): " +
                        "Programming error: View is not Checkable."
                );
            }

            if (!(view instanceof IObserverListener))
            {
                throw new RuntimeException(
                    "ViewGroupSingleTargetSelector::addSingleTargetSelection(): " +
                        "Programming error: View is not ObserverListener."
                );
            }

            int i = childIdx;
            ((IObserverListener) view).addOnClickObserver(() -> {
                ((Checkable) view).setChecked(true);
                for (int j = 0; j < viewGroup.getChildCount(); ++j)
                {
                    if (i != j)
                    {
                        View tmp = viewGroup.getChildAt(j);
                        ((Checkable) tmp).setChecked(false);
                    }
                }
            });
        }
    }
}

File: app/src/main/java/com/liweiyap/narradir/ui/ViewGroupSingleTargetSelector.kt
package com.liweiyap.narradir.ui

import android.view.View
import android.view.ViewGroup
import android.widget.Checkable

object ViewGroupSingleTargetSelector {
    @Throws(RuntimeException::class)
    fun addSingleTargetSelection(viewGroup: ViewGroup?) {
        if (viewGroup == null) {
            return
        }

        for (childIdx: Int in 0 until viewGroup.childCount) {
            val view: View = viewGroup.getChildAt(childIdx)

            if (view !is Checkable) {
                throw RuntimeException(
                    "ViewGroupSingleTargetSelector::addSingleTargetSelection(): " +
                            "Programming error: View is not Checkable.")
            }

            if (view !is IObserverListener) {
                throw RuntimeException(
                    "ViewGroupSingleTargetSelector::addSingleTargetSelection(): " +
                            "Programming error: View is not IObserverListener.")
            }

            (view as IObserverListener).addOnClickObserver {
                (view as Checkable).isChecked = true
                for (idx: Int in 0 until viewGroup.childCount) {
                    if (childIdx != idx) {
                        val tmp: View = viewGroup.getChildAt(idx)
                        (tmp as Checkable).isChecked = false
                    }
                }
            }
        }
    }
}